File: src/services/remote/esri.ts
// Generic ESRI FeatureServer query helper.
//
// App-unused since the migration to pre-built static artifacts:
// runtime data comes from tools/data-build outputs. Kept for future
// optional ESRI overlay layers (docs/07, ESRI REST section).
//
// Asks for f=geojson so the server emits standard GeoJSON
// (RFC 7946) directly. ESRI handles ring orientation and
// MultiPolygon assembly server-side. No client-side ring grouping
// or geometry conversion needed.
//
// CRS: outSR=4326 pins WGS84 in the response. Stores always hold
// WGS84 per docs/07 "CRS handling".

import type { Feature, FeatureCollection, Geometry } from 'geojson'

// ESRI extends the GeoJSON FeatureCollection with a top-level
// properties bag carrying exceededTransferLimit. Standard FeatureCollection
// has no such field, so this is an ESRI-specific shape.
type EsriGeoJsonResponse = FeatureCollection & {
  properties?: { exceededTransferLimit?: boolean }
  error?: { code?: number; message: string }
}

export type QueryOptions = {
  where?: string
  outFields?: string[]
  returnGeometry?: boolean
  pageSize?: number
  onPageComplete?: (completed: number, totalPages: number) => void
}

export async function fetchLayerLastEditDate(layerUrl: string): Promise<number | null> {
  const url = `${layerUrl}?f=json`
  const response = await fetch(url)
  if (!response.ok) {
    throw new Error(
      `ESRI layer metadata fetch failed: ${response.status} ${response.statusText} (${url})`,
    )
  }
  const data = (await response.json()) as {
    editingInfo?: { lastEditDate?: number }
    error?: { message: string }
  }
  if (data.error) {
    throw new Error(`ESRI layer metadata error: ${data.error.message}`)
  }
  return data.editingInfo?.lastEditDate ?? null
}

async function queryCount(layerUrl: string, where: string): Promise<number> {
  const params = new URLSearchParams({
    where,
    returnCountOnly: 'true',
    f: 'json',
  })
  const url = `${layerUrl}/query?${params}`
  const response = await fetch(url)
  if (!response.ok) {
    throw new Error(
      `ESRI count query failed: ${response.status} ${response.statusText} (${url})`,
    )
  }
  const data = (await response.json()) as { count?: number; error?: { message: string } }
  if (data.error) {
    throw new Error(`ESRI count query error: ${data.error.message}`)
  }
  if (typeof data.count !== 'number') {
    throw new Error(`ESRI count query returned no count (${url})`)
  }
  return data.count
}

async function fetchPage<G extends Geometry>(
  layerUrl: string,
  where: string,
  outFields: string[],
  returnGeometry: boolean,
  pageSize: number,
  resultOffset: number,
): Promise<Feature<G>[]> {
  const params = new URLSearchParams({
    where,
    outFields: outFields.join(','),
    returnGeometry: String(returnGeometry),
    outSR: '4326',
    f: 'geojson',
    resultOffset: String(resultOffset),
    resultRecordCount: String(pageSize),
  })

  const url = `${layerUrl}/query?${params}`
  const response = await fetch(url)
  if (!response.ok) {
    throw new Error(
      `ESRI query failed: ${response.status} ${response.statusText} (${url})`,
    )
  }
  const data = (await response.json()) as EsriGeoJsonResponse
  if (data.error) {
    throw new Error(`ESRI query error: ${data.error.message}`)
  }
  return data.features as Feature<G>[]
}

// Pulls every feature matching the query. Fetches the count first,
// then fires all page requests in parallel. Page size of 2000
// matches the FeatureServer maxRecordCount default.
//
// Generic G narrows the returned geometry type so callers consume
// `Feature<Polygon | MultiPolygon>[]` (or whatever shape they expect)
// without casting.
export async function queryAllFeatures<G extends Geometry = Geometry>(
  layerUrl: string,
  options: QueryOptions = {},
): Promise<Feature<G>[]> {
  const {
    where = '1=1',
    outFields = ['*'],
    returnGeometry = true,
    pageSize = 2000,
    onPageComplete,
  } = options

  const total = await queryCount(layerUrl, where)
  if (total === 0) return []

  const offsets: number[] = []
  for (let offset = 0; offset < total; offset += pageSize) {
    offsets.push(offset)
  }

  let completed = 0
  const pages = await Promise.all(
    offsets.map(async (offset) => {
      const page = await fetchPage<G>(layerUrl, where, outFields, returnGeometry, pageSize, offset)
      completed++
      onPageComplete?.(completed, offsets.length)
      return page
    }),
  )

  return pages.flat()
}

File: tools/data-build/ingest/esri.py
"""ESRI FeatureServer fetch helper.

Python port of src/services/remote/esri.ts. Count-first parallel page
fetch, GeoJSON response, WGS84 output.
"""

from __future__ import annotations

from concurrent.futures import ThreadPoolExecutor, as_completed

import geopandas as gpd
import requests


class EsriFetchError(Exception):
    """Raised when an ESRI FeatureServer request fails."""
    pass


def _check_esri_error(data: dict, context: str) -> None:
    if "error" in data:
        code = data["error"].get("code", "")
        msg = data["error"].get("message", "unknown error")
        raise EsriFetchError(f"{context}: {msg} (code {code})")


def fetch_count(layer_url: str, where: str = "1=1") -> int:
    params = {"where": where, "returnCountOnly": "true", "f": "json"}
    try:
        resp = requests.get(f"{layer_url}/query", params=params, timeout=30)
        resp.raise_for_status()
    except requests.RequestException as e:
        raise EsriFetchError(f"Count query failed for {layer_url}: {e}") from e
    data = resp.json()
    _check_esri_error(data, f"Count query for {layer_url}")
    if "count" not in data:
        raise EsriFetchError(f"Count query returned no count for {layer_url}")
    return data["count"]


def fetch_page(
    layer_url: str,
    where: str,
    out_fields: list[str],
    offset: int,
    page_size: int,
) -> dict:
    params = {
        "where": where,
        "outFields": ",".join(out_fields),
        "returnGeometry": "true",
        "outSR": "4326",
        "f": "geojson",
        "resultOffset": str(offset),
        "resultRecordCount": str(page_size),
    }
    try:
        resp = requests.get(f"{layer_url}/query", params=params, timeout=60)
        resp.raise_for_status()
    except requests.RequestException as e:
        raise EsriFetchError(f"Page fetch failed at offset {offset}: {e}") from e
    data = resp.json()
    _check_esri_error(data, f"Page fetch at offset {offset}")
    return data


def fetch_all_features(
    layer_url: str,
    out_fields: list[str] | None = None,
    where: str = "1=1",
    page_size: int = 2000,
) -> gpd.GeoDataFrame:
    """Fetch all features from an ESRI FeatureServer layer.

    Returns a GeoDataFrame in EPSG:4326. Raises EsriFetchError on
    network or ESRI API errors.
    """
    if out_fields is None:
        out_fields = ["*"]

    total = fetch_count(layer_url, where)
    if total == 0:
        return gpd.GeoDataFrame()

    print(f"  Fetching {total:,} features from {layer_url}")
    offsets = list(range(0, total, page_size))
    all_features: list[dict] = []

    with ThreadPoolExecutor(max_workers=6) as pool:
        futures = {
            pool.submit(fetch_page, layer_url, where, out_fields, off, page_size): off
            for off in offsets
        }
        for i, future in enumerate(as_completed(futures), 1):
            page = future.result()
            all_features.extend(page.get("features", []))
            print(f"    Page {i}/{len(offsets)} ({len(all_features):,} features)")

    geojson = {"type": "FeatureCollection", "features": all_features}
    gdf = gpd.GeoDataFrame.from_features(geojson, crs="EPSG:4326")
    print(f"  Fetched {len(gdf):,} features")
    return gdf
